reconcile/test/test_slack_base.py

```diff
@@ -0,0 +1,140 @@
+import pytest
+
+from reconcile.slack_base import slackapi_from_slack_workspace, \
+    slackapi_from_permissions
+
+
+def create_api_config():
+    return {
+        "global": {
+            "max_retries": 5,
+            "timeout": 30
+        },
+        "methods": [
+            {
+                "name": "users.list",
+                "args": "{\"limit\":123}"
+            },
+        ]
+    }
+
+
+@pytest.fixture
+def slack_workspace():
+    return {
+        "workspace": {
+            "name": "coreos",
+            "integrations": [
+                {
+                    "name": "dummy",
+                    "token": {
+                        "path": "some/path",
+                        "field": "bot_token",
+                    },
+                    "channel": "test",
+                    "icon_emoji": "test_emoji",
+                    "username": "foo"
+                },
+            ],
+            "api_client": create_api_config(),
+        }
+    }
+
+
+@pytest.fixture
+def unleash_slack_workspace():
+    return {
+        "workspace": {
+            "name": "coreos",
+            "integrations": [
+                {
+                    "name": "unleash-watcher",
+                    "token": {
+                        "path": "some/path",
+                        "field": "bot_token",
+                    }
+                },
+            ],
+            "api_client": create_api_config(),
+        },
+        "channel": "test",
+        "icon_emoji": "unleash",
+        "username": "test"
+    }
+
+
+@pytest.fixture
+def permissions_workspace():
+    return {
+        "workspace": {
+            "name": "coreos",
+            "token": {
+                "path": "some/path",
+                "field": "bot_token",
+            },
+            "api_client": create_api_config(),
+            "managedUsergroups": [
+                "foo"
+            ]
+        }
+    }
+
+
+@pytest.fixture
+def slack_base(mocker):
```

Comment:
I missed that this was changed. I think there are at least a few approaches to taking care of patching these dependencies in `__init__`:

1. A single patch per fixture, which is very explicit and simple to follow if named well (ex. `patch_secret_reader_read` and `patch_initiate_usergroups`
2. Use xunit style setup/teardown - the setup is now implicit, which might not be as easy to read from within the test, but at least it's a standard convention and can be easily referenced from classes using `self`
3. Doing a lot of setup in a fixture - my biggest concern here is you get the implicit nature of xunit setup, but also have to start doing things like returning tuples to pass the mocked objects. You'll end up with `ValueError` if the fixture ever changes to return more objects.

My general preference is 1 and I'm okay with 2. The third option I'm a little hesitant with unless it's named and documented that people won't use it for a purpose other than it was intended. In this case, may `patch_slack_api_init_deps` and a docstring indicating that this is only meant to patch those dependencies that SlackApi is creating in `__init__` that need to be patched to be able to unit test the class. You might also consider using a NamedTuple or dataclass if you feel strongly about doing it this way. That might help reduce the need to refactor tests in the fixture changes, and reduce the chance that someone pulls the mocks in the wrong order, etc.

Reply:
Yeah, I saw that in slack_api and thought that's the way to go :) 
I'll create two individual fixtures.

Reply:
The difference with the `slack_api` fixture is that it's actually returning an object more than it's intended to just patch objects. It's unlikely that fixture ends up becoming very general use (although a docstring could make the intent clearer for `slack_api`). Also, it uses namedtuple to expose everything, so adding a new mock wouldn't require refactoring all of the tests that don't care about that new mock. Hopefully this reasoning makes sense. I'm just really trying to avoid very general fixtures that end up being used for different use cases and are changed as a result, which could affect existing tests.